File: 04-CoffeeMachine/src/main/java/coffee/machine/domain/CoffeeMachineLogic.java
package coffee.machine.domain;

import java.util.HashMap;

public class CoffeeMachineLogic implements IValidateDrinkInstruction {
    final HashMap<KindOfDrink, Double> drinkPrices = new HashMap<>();
    private final BeverageQuantityChecker beverageQuantityChecker;
    private final EmailNotifier emailNotifier;
    private final FinancialReport financialReport;


    public CoffeeMachineLogic(BeverageQuantityChecker beverageQuantityChecker, EmailNotifier emailNotifier, FinancialReport businessReport) {
        this.beverageQuantityChecker = beverageQuantityChecker;
        this.emailNotifier = emailNotifier;
        this.financialReport = businessReport;

        this.drinkPrices.put(KindOfDrink.Tea, 0.4);
        this.drinkPrices.put(KindOfDrink.Coffee, 0.6);
        this.drinkPrices.put(KindOfDrink.Chocolate, 0.5);
        this.drinkPrices.put(KindOfDrink.OrangeJuice, 0.6);
    }

    public DrinkInstruction checkValidityForDrinkPreparation(DrinkInstruction drinkInstruction) {

        if (this.isThereShortageForThisDrink(drinkInstruction)) {
            return this.drinkInstructionFailedForShortage(drinkInstruction);
        }

        if (!drinkInstruction.isCustomerHaveEnoughMoney(this.drinkPrices)) {
            return this.drinkInstructionFailedForMissingMoney(drinkInstruction);
        }

        if (this.isFinancialReportIsNeeded()) {
            this.financialReport.addBeverageSold(drinkInstruction, this.drinkPrices);
        }

        return drinkInstruction;
    }

    private boolean isThereShortageForThisDrink(DrinkInstruction drinkInstruction) {
        return beverageQuantityChecker != null && beverageQuantityChecker.isEmpty(drinkInstruction.getDrink().toString());
    }

    private DrinkInstructionFailed drinkInstructionFailedForShortage(DrinkInstruction instructions) {
        if (this.emailNotifier != null) {
            this.emailNotifier.notifyMissingDrink(instructions.getDrink().toString());
        }
        return new DrinkInstructionFailed(String.format("M:%s shortage (a notification has been sent to our logistic division). Please pick another option.", instructions.getDrink().toString()));
    }

    private boolean isFinancialReportIsNeeded() {
        return this.financialReport != null;
    }

    private DrinkInstruction drinkInstructionFailedForMissingMoney(DrinkInstruction instruction) {
        double moneyMissing = drinkPrices.get(instruction.drink) - instruction.customerMoney;
        return new DrinkInstructionFailed(String.format("M:Missing %.2f euro", moneyMissing));
    }
}

File: 04-CoffeeMachine/version-csharp-thomas/CoffeeMachine.Tests/Domain/CoffeeMachineLogic.cs
namespace CoffeeMachine.Tests.Domain
{
    public class CoffeeMachineLogic
    {
        private readonly ITalkToTheDrinkMaker _drinkMakerAdapter;
        private readonly ICheckBeverageQuantity _beverageQuantityChecker;
        private readonly INotifyViaEMail _emailNotifier;
        private readonly IForwardMessagesToEndUser _messageToEndUserForwarder;
        private readonly CoffeeMachineAnalytics _analytics;
        
        private decimal? _receivedMoney;
        private CustomerIncomingOrder _receivedOrder;

        public CoffeeMachineLogic(ITalkToTheDrinkMaker drinkMakerAdapter, ICheckBeverageQuantity beverageQuantityChecker, INotifyViaEMail emailNotifier)
        {
            _drinkMakerAdapter = drinkMakerAdapter;
            _beverageQuantityChecker = beverageQuantityChecker;
            _emailNotifier = emailNotifier;
            _analytics = new CoffeeMachineAnalytics();
            _messageToEndUserForwarder = new MessageToEndUserForwarder(drinkMakerAdapter);
        }
        
        private decimal ReceivedMoney => _receivedMoney.HasValue ? _receivedMoney.Value : 0;

        public void Receive(CustomerIncomingOrder order)
        {
            if (ShortageIsDetectedForThisBeverage(order))
            {
                _messageToEndUserForwarder.SendMessage($"{order.Product.ToString()} shortage (a notification has been sent to our logistic division). Please pick another option.");
                _emailNotifier.NotifyMissingDrink(order.Product.ToString());
                return;
            }

            ReplacePreviousOrderWithNewOne(order);

            if (!ReceivedEnoughMoneyFor(order))
            {
                ComputeMissingAmountAndSendMessageForIt(order);
                return;
            }

            var instructions = DrinkMakerAdapter.AdaptToDrinkMakerInstruction(order);
            MakeBeverageTransaction(instructions);
        }

        public void ReceiveMoney(decimal amountInEuro)
        {
            CollectMoney(amountInEuro);

            if (ReceivedEnoughMoneyFor(_receivedOrder))
            {
                var instruction = DrinkMakerAdapter.AdaptToDrinkMakerInstruction(_receivedOrder);

                MakeBeverageTransaction(instruction);
            }
            else
            {
                ComputeMissingAmountAndSendMessageForIt(_receivedOrder);
            }
        }

        public FinancialReport GetFinancialReport()
        {
            return _analytics.GenerateReport();
        }

        private bool ShortageIsDetectedForThisBeverage(CustomerIncomingOrder order)
        {
            return _beverageQuantityChecker.IsEmpty(order.Product.ToString());
        }

        private void ReplacePreviousOrderWithNewOne(CustomerIncomingOrder order)
        {
            _receivedOrder = order;
        }

        private void MakeBeverageTransaction(string instructions)
        {
            _drinkMakerAdapter.Send(instructions);
            
            // Log the transaction
            _analytics.Record(_receivedOrder, ReceivedMoney);

            // close the transaction
            _receivedOrder = null;
            _receivedMoney = null;
        }

        private void ComputeMissingAmountAndSendMessageForIt(CustomerIncomingOrder order)
        {
            if (order == null)
            {
                _messageToEndUserForwarder.SendMessage($"Pick a beverage");
                return;
            }

            var missingAmount = Prices.ComputeMissingAmount(order.Product, ReceivedMoney);

            _messageToEndUserForwarder.SendMissingAmountMessage(missingAmount);
        }

        private void CollectMoney(decimal amountInEuro)
        {
            _receivedMoney = _receivedMoney + amountInEuro ?? amountInEuro;
        }

        private bool ReceivedEnoughMoneyFor(CustomerIncomingOrder order)
        {
            if (order == null)
            {
                return false;
            }

            return (ReceivedMoney >= Prices.GetUnitPriceFor(order.Product));
        }
    }
}